nvse/nvse/ScriptAnalyzer.cpp

```diff
@@ -1267,39 +1268,57 @@ std::string ScriptParsing::ScriptAnalyzer::DecompileScript()
 		ReplaceAll(nextLine, "\n", '\n' + tabStr);
 		scriptText += tabStr;
 
-		if (!lastOpcode && !script->varList.Empty())
-		{
-			if (opcode == static_cast<UInt16>(ScriptStatementCode::ScriptName))
-				scriptText += nextLine + '\n' + '\n';
-			for (auto* var : script->varList)
-			{
-				if (!var)
-					continue;
-				ScriptVariableToken token(script, ExpressionCode::None, var, nullptr);
-				if (arrayVariables.contains(var))
-					scriptText += "Array_var ";
-				else if (stringVariables.contains(var))
-					scriptText += "String_var ";
-				else if (var->IsReferenceType(script))
-					scriptText += "Ref ";
-				else if (var->type == Script::VariableType::eVarType_Float)
-					scriptText += "Float ";
-				else
-					scriptText += "Int ";
-				scriptText += token.ToString() + '\n';
-			}
-			scriptText += '\n';
-			if (opcode != static_cast<UInt16>(ScriptStatementCode::ScriptName))
-				scriptText += nextLine + '\n';
+		if (opcode == static_cast<UInt16>(ScriptStatementCode::ScriptName)) {
+			scriptHeader += nextLine + "\n\n";
+		}
+		else {
+			scriptText += nextLine + '\n';
 		}
-		else scriptText += nextLine + '\n';
 
 		if (isNeutral || isAdd)
 			++numTabs;
 		
 		lastOpcode = opcode;
 	}
-	return scriptText;
+
+	// Build header string
+	if (!isLambdaScript) {
+		std::unordered_set<std::string> addedVars{};
+		auto addVar = [&] (VariableInfo *var) {
+			const auto name = std::string(var->name.CStr());
+			if (!var || addedVars.contains(name)) {
+				return;
+			}
+			addedVars.insert(name);
+				
+			ScriptVariableToken token(script, ExpressionCode::None, var, nullptr);
+			if (arrayVariables.contains(var))
+				scriptHeader += "Array_var ";
+			else if (stringVariables.contains(var))
+				scriptHeader += "String_var ";
+			else if (var->IsReferenceType(script))
+				scriptHeader += "Ref ";
+			else if (var->type == Script::VariableType::eVarType_Float)
+				scriptHeader += "Float ";
+			else
+				scriptHeader += "Int ";
+			scriptHeader += token.ToString() + '\n';
+		};
+	
+		for (auto* var : script->varList) {
+			addVar(var);
+		}
+
+		for (auto* var : stringVariables) {

```

Comment:
Correct me if I'm wrong but strings and arrays are also a part of the var list, won't this cause duplicate definitions?

Reply:
Oh I see you added a set. What's the basis for this refactoring?

Reply:
Vars were getting duplicated with lambdas / messing with decompiler output

```
Array_var mpNameToEditorId
Int iter
Int edid
Int qname
Array_var tracked
Int i
String_var questPath
String_var questName
String_var editorId

Begin Function { }
	If eval GetActiveMenuMode != 1023
		Return
	EndIf

	Let mpNameToEditorId := ar_Construct ("stringmap")
	ar_ForEach GetQuests Array_var mpNameToEditorId
	Array_var iter
	String_var edid
	String_var qname
	Int tracked
	Int i
	Int questPath
	Int questName
	Int editorId
	
	Begin Function { }
		Let edid := (GetEditorID (iter[0]))
		Let qname := $iter[0]
		Let mpNameToEditorId[qname] := edid
	End
```

Reply:
I see